# Zero division error

In [3]:
a = 10 
try:
    print(a/0)
except ZeroDivisionError:
    print("Divison by zero is not allowed")

print(f"Zero division error occured for dividing {a} by 0")

Divison by zero is not allowed
Zero division error occured for dividing 10 by 0


# Index Error 

In [5]:
l = [1,2,3,4]
try:
    print(l[7])
except IndexError:
    print("Index out of range error")
print("Index out of range error occured for accessing index 7 in list l")

Index out of range error
Index out of range error occured for accessing index 7 in list l


### Write a python code to run a certain block of code if the code block inside try runs without any errors. 

In [7]:
a = 10 
try:
    print(a/10)
except ZeroDivisionError:
    print("Zero division error has occured")
else:
    print("No error has occured")

1.0
No error has occured


### Write a python code to run a certain block of code no matter whether there is an exception or not.

In [8]:
l = [1,2,3,4,5]
try:
    print(l[2])
except IndexError:
    print("Index out of range error")
else:
    print("No error has occured")
finally:
    print("Error hok ki na hok ami print hoboi")

3
No error has occured
Error hok ki na hok ami print hoboi


In [9]:
a = 10
try:
    print(a/0)
except ZeroDivisionError:
    print("Division by zero is not allowed")
else:
    print("No error has occured")
finally:
    print("Hello happy rainy day!")

Division by zero is not allowed
Hello happy rainy day!


##### Write a python code to raise Exceptions (Python's built-in Exception objects) by yourself to indicate that something went wrong

In [12]:
a = 10 
b = 5
if(b==0):
    raise ZeroDivisionError("Division by zero is not allowed")
else:
    print(a/b)


2.0


In [13]:
a = 10 
if(a<0):
    raise ValueError("Negative value is not allowed")
else:
    print(a**0.5)

3.1622776601683795


In [16]:
l = [1,2,3,4]
length = len(l)
a = 10
if(a>length):
    raise Exception("Index out of range")
else:
    print(l[a])


Exception: Index out of range

#### Write  a  python  code  to  define  custom  exceptions  by  creating  a  new  class  that  is  derived  from  the  built-in exception class

In [18]:
class CustomError(Exception):
    pass 

In [19]:
a = 10 
b = 'abc'

if(type(b) is not int):
    raise CustomError("Addition can't be performed between a integer and a string")
else:
    print(a+b)

CustomError: Addition can't be performed between a integer and a string

#### You're going to write an interactive calculator! User input is assumed to be a formula that consist of a number, an operator (at least + and -), and another number, separated by white space (e.g. 1 + 1). Split user input using str.split(), and check whether the resulting list is valid:  ✓ If the input does not consist of 3 elements, raise a FormulaError, which is a custom Exception.  ✓ Try  to  convert  the  first  and  third  input  to  a float (like  so: float_value  =  float(str_value)).  Catch any ValueError that occurs, and instead raise a FormulaError  ✓ If the second input is not '+' or '-', again raise a FormulaError  If the input is valid, perform the calculation and print out the result. The user is then prompted to provide new input, and so on, until the user enters quit.  An interaction could look like this:  >>> 1 + 1  2.0  >>> 3.2 - 1.5  1.7000000000000002  >>> quit 
 